File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def dfs(node, nodes)
  return if node == nil
  dfs(node.left, nodes)
  nodes << node
  dfs(node.right, nodes)
end

def recover_tree(root)
  nodes = []
  dfs(root, nodes)
  vals = nodes.map{ |node| node.val }
  vals.sort!
  nodes.each_with_index{ |node, i| node.val = vals[i] }
end

File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *   Val int
 *   Left *TreeNode
 *   Right *TreeNode
 * }
 */
func recoverTree(root *TreeNode)  {
  nodes := []*TreeNode{}
  dfs(root, &nodes)
  vals := []int{}
  for _, node := range nodes {
    vals = append(vals, node.Val)
  }
  sort.Ints(vals)
  for i := 0; i < len(nodes); i++ {
    nodes[i].Val = vals[i]
  }
}

func dfs(node *TreeNode, nodes *[]*TreeNode) {
  if node == nil {
    return
  }
  dfs((*node).Left, nodes)
  *nodes = append(*nodes, node)
  dfs((*node).Right, nodes)
}